spring-cloud-gcp-pubsub-stream-binder/src/main/java/org/springframework/cloud/gcp/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java

```diff
@@ -58,24 +68,51 @@ public ProducerDestination provisionProducerDestination(String name,
 	public ConsumerDestination provisionConsumerDestination(String name, String group,
 			ExtendedConsumerProperties<PubSubConsumerProperties> properties)
 			throws ProvisioningException {
-		// Use group name as subscription name
+		String subscription;
+		// Use <topic>.<groupName> as subscription name
 		// Generate anonymous random group, if one not provided
-		String subscription = !StringUtils.hasText(group) ?
-				"anonymous." + name + "." + UUID.randomUUID().toString()
-				: name + "." + group;
+		boolean anonymous = false;
+		if (StringUtils.hasText(group)) {
+			subscription = name + "." + group;
+		}
+		else {
+			subscription = "anonymous." + name + "." + UUID.randomUUID().toString();
+			anonymous = true;
+		}
 
-		if (this.pubSubAdmin.getSubscription(subscription) == null) {
+		Subscription pubSubSubscription = this.pubSubAdmin.getSubscription(subscription);
+		if (pubSubSubscription == null) {
 			if (properties.getExtension().isAutoCreateResources()) {
 				if (this.pubSubAdmin.getTopic(name) == null) {
 					this.pubSubAdmin.createTopic(name);
 				}
 
 				this.pubSubAdmin.createSubscription(subscription, name);
+
+				if (anonymous) {
+					this.anonymousSubscriptions.add(subscription);
+				}
 			}
 			else {
 				throw new ProvisioningException("Non-existing '" + subscription + "' subscription.");
 			}
 		}
+		else if (!pubSubSubscription.getTopic().equals(name)) {
+			throw new ProvisioningException("Existing '" + subscription + "' subscription is for a different topic '"
+					+ pubSubSubscription.getTopic() + "'.");
+		}
 		return new PubSubConsumerDestination(subscription);
 	}
+
+	public void afterUnbindConsumer(ConsumerDestination destination) {
+		if (this.anonymousSubscriptions.contains(destination.getName())) {
```

Comment:
I think you just can use and atomic operation here: `this.anonymousSubscriptions.remove(destination.getName()) != null`.